Added "probe" support to project and container resources

Dona278 opened the pull request:
- Allow add a "probe" to a project, container or executable resource: at this moment only http probes are managed;
- Added support for "probe" to kubernetes, azure container app and azure app service. docker-compose allow only one probe called "HealthCheck" and seems to works only with command, so for now it was ignored until it will be more clear how to integrate.
- Added tests;

Fixes #890 

## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
  - [ ] No
- Did you add public API?
  - [x] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [x] Yes
      - [ ] No
  - [ ] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change require an update in our Aspire docs?
  - [x] Yes
  - [ ] No

Dona278 committed:
Added "probe" support to project and container resources

- Allow add a "probe" to a project or container resource: at this moment only http probes are managed;
- Added support for "probe" to kubernetes, azure container app and docker-compose publisher;
- Added tests;

davidfowl commented:
ExecutableResource also needs IResourceWithProbes.

davidfowl commented on src/Aspire.Hosting.Azure.AppContainers/ContainerAppContext.cs:
```diff
@@ -815,13 +816,51 @@ private void AddContainerRegistryParameters(ContainerAppConfiguration app)
+                    HttpGet = new()
+                    {
+                        Path = endpointProbeAnnotation.Path,
+                        Port = endpointProbeAnnotation.EndpointReference.TargetPort.Value,
```
We don't want to resolve the target port this way. See https://github.com/dotnet/aspire/blob/f2acbde213f728abb956a1c2adde4120c42be86b/src/Aspire.Hosting.Azure.AppContainers/ContainerAppContext.cs#L496C32-L496C40 for more details on how we do this today for other endpoints.

Dona278 replied:
GetValue() works with BicepValue<string> but probe require int. Can I simple parse string into int or there is something with parameter allocation that I'll break?

Dona278 committed:
- Added `IResourceWithProbes` to `ExecutableResource`;
- Used processed endpoint mappings to resolve target port for probe in `ContainerAppContext`;
- Removed "probes" from docker-compose because its unclear how to do translation in a way that always works;

Dona278 committed:
- Added more properties to probe annotation;
- Changed how kubernetes publisher resolve port for probe;

davidfowl commented on tests/Aspire.Hosting.Docker.Tests/Snapshots/DockerComposePublisherTests.ResourceWithProbes.verified.yaml:
```diff
@@ -0,0 +1,29 @@
+﻿services:
```
This should be deleted right?

Dona278 commented on src/Aspire.Hosting/ResourceBuilderExtensions.cs:
```diff
@@ -2259,4 +2259,56 @@ public static IResourceBuilder<T> WithComputeEnvironment<T>(this IResourceBuilde
+    /// <returns>A reference to the <see cref="IResourceBuilder{T}"/>.</returns>
+    private static IResourceBuilder<T> WithProbe<T>(this IResourceBuilder<T> builder, ProbeAnnotation probeAnnotation) where T : IResourceWithProbes
+    {
+        if (builder.Resource.Annotations.OfType<ProbeAnnotation>().Any(e => e.Type == probeAnnotation.Type))
+        {
+            throw new DistributedApplicationException($"A probe with type '{probeAnnotation.Type}' already exists");
+        }
```
There is no targer port in that model. I found [here](https://github.com/dotnet/aspire/blob/6cbacd2c2564d59f285ac3ddf588fe198e006e94/src/Aspire.Hosting.Kubernetes/KubernetesServiceResourceExtensions.cs#L96) that "port" is used also for "target port".

Dona278 replied:
Ok I'll change the behavior to "replace".
But this is a replace by filter for a specific property other then the annotation type, and current implementation of `WithAnnotation` check only for same annotation type. 
I though for a new overload like:
```cs
public IResourceBuilder<T> WithAnnotation<TAnnotation>(TAnnotation annotation, ResourceAnnotationMutationBehavior behavior = ResourceAnnotationMutationBehavior.Append, Func<TAnnotation, bool>? filter = null) where TAnnotation : IResourceAnnotation;
```
but make sense only for `ResourceAnnotationMutationBehavior.Replace`. Honeslty I don't want to copy paste code and I think that can be useful for other in the future no? How do you want to do? Implement an annotation replace within `WithProbe` or add a `ReplaceAnnotation` which is different from `WithAnnotation` because allow filter? And maybe `ReplaceAnnotation` can be called within `WithAnnotation` as well when needed. Or what else?

Dona278 replied:
Ok I'll change the behavior to "replace".
Side note: this is a replace by filter for a specific property other then annotation type. Current implementation of `WithAnnotation` check only for same annotation type. `WithCommand` have the same requirement. Why we have only `WithAnnotation` with replace behavior and not a `ReplaceAnnotation`? Or better, we can have a `ReplaceAnnotation` which is different from `WithAnnotation` because allow filter and maybe `ReplaceAnnotation` is called within `WithAnnotation` when needed?

Dona278 committed:
Delete DockerComposePublisherTests.ResourceWithProbes.verified.yaml

Dona278 commented on src/Aspire.Hosting.Kubernetes/Extensions/ResourceExtensions.cs:
```diff
@@ -343,6 +344,58 @@ private static ContainerV1 WithContainerSecrets(this ContainerV1 container, Kube
+                    HttpGet = new()
+                    {
+                        Path = endpointProbeAnnotation.Path,
+                        Port = endpointMapping.Port,
```
Yeah is what I linked in the comment. That method use port in all cases. You mean "Use this method however"? In that case we need to move to another place (is static and private) but since it is called from one file (where is placed) from another extension which is called only from "KubernetesResource.cs", I suggest to move all methods of `KubernetesServiceResourceExtensions.cs` within `KubernetesResource.cs` and in this way I can use `GetEndpointValue`

davidfowl replied:
Use the method anyways so that we have a single source of truth for bug fixes 😄

Dona278 committed:
Moved content of `KubernetesServiceResourceExtensions.cs` into `KubernetesResource.cs` (which is the only one that use it)

- Handled probes in two steps: process and assign (as done for endpoints and volumes);

davidfowl commented:
App Service also supports health checks it seems https://medium.com/@mariekie/bicep-enable-azure-health-checks-on-app-services-dbc057abed35

Dona278 commented:
> App Service also supports health checks it seems https://medium.com/@mariekie/bicep-enable-azure-health-checks-on-app-services-dbc057abed35

There is nothing other than path, so how we want to translate a list of probes? Take the first one (major priority for liveness) and ignore other properties?

Dona278 committed:
Added "probe" support to "AzureAppService";

davidfowl commented:
I don’t like the idea that if generic metadata doesn’t work everywhere that we don’t have it. I think health probes are a well understood concept and we should model them in aspire. It’s a valid question what to do with the in various environments and you can always override the with provider specific behavior.

davidfowl commented:
We need an HttpProbeHealthCheck now 😄

Dona278 commented:
> We need an HttpProbeHealthCheck now 😄

You mean "WithHttpProbeHealthCheck" right? 😄 
However I'll change the signature of the method to use nulls but for path I think that as default value is better "/health" which is the one used in `ServiceDefaults.MapDefaultEndpoints()`

Dona278 committed:
- Updated `WithHttpProbe` method signature (remove defaults);

- Aligned endpoint selector with `WithHttpHealthCheck` method;

Dona278 committed:
Marked "IResourceWithProbe" as experimental

Co-authored-by: Safia Abdalla <safia@safia.rocks>

Dona278 committed:
- Managed experimental attribute;

- Updated code in `ContainerAppContext.cs` for probe as we did in `KebernetesResource.cs`;
- Changed behavior on how probe annotation is added to a resource: now replace an existing of the same type. Updated test;

Dona278 committed:
Implicit call `WithHttpHealthCheck` within `WithHttpProbe`. Updated method comment;

Dona278 committed:
Merge branch 'main' into probes

Dona278 committed:
Fixed merge branch error

davidfowl commented:
@DamianEdwards @mitchdenny this is good for another set of testing!

davidfowl commented:
@Dona278 have you tested this all the way through to deployment? I'm asking because I'm wondering if:

```bicep
{
              failureThreshold: 3
              httpGet: {
                path: '/alive'
                port: int(webfrontend_containerport)
                scheme: 'HTTPS'
              }
              initialDelaySeconds: 5
              periodSeconds: 5
              successThreshold: 1
              timeoutSeconds: 1
              type: 'Readiness'
            }
```

Is hitting the internal container port and if the scheme https is correct.

Dona278 committed:
Use scheme property of endpoint mapping instead of the endpoint reference one (from annotation)

Dona278 commented:
I played a bit with endpoints (which are the source of the issue with probes) and with deployment (ACA):

Recap:
All works good except for probe which is published as "https", but this isn't an error or a hardcoded scheme, it is read from all endpoints associated to the resource, so why the service listen over "http" but probe is "https"?
I found that within `AddProject()` the call to `WithProjectDefaults()` in `PublishMode` both "http" and "https" are added by default (auto-binding) if the project is a "web project", and "https" is marked as `DefaultHttpsEndpoint=true`.
So the resource really has "https" endpoint but, in the case of ACA, the port used by service isn't managed within publisher based on endpoints, it came from env var "HTTP_PORTS" which was added by `SetBothPortsEnvVariables` (called within `WithProjectDefaults()` in `PublishMode`) that SKIP the "https" endpoint because added "as default".
So first Aspire adds two endpoints and then ignore "https" one when need to provide env var ports.

HttpProbes, as HttpHealthChecks, use a common endpoint selector which as default prioritize "https" if exists, but "HealthCheck" isn't a published entity so the issue never comes out, but if you check annotations during publish / deploy "HealthCheck" has the same issue.

Finally, in ACA, the "https" autogenerated endpoint doesn't create issue because is added after "http" and both doesn't have port or target port set, and during process the endpoints are grouped by port (so null) and the first group will be used as ingress and without port a default "80" will be used and the scheme will be set based on port (so http).
Again, the endpoint is present but ignored due to order....

If the resource came with a "WithHttpEndpoint" or "WithHttpsEndpoint" there are no issues because endpoints are manually addeed.

Now:
I was trying to correctly manage this behavior in ACA (and then in other publishers) but I think that isn't the best way copy-paste logic among publishers by using the same logic of env var, so replace (we cannot remove) if `DefaultHttpsEndpoint=true`.
First because if someone forgot this in a new publisher, then the publisher has an issue.
Second because I'm not a fan of  "silent change a behavior (https -> http)" , but this is only for default case so....

I was wondering if it Isn't better to manage this in "WithHttpHealthCheck" and "WithHttpProbe" in `PublishMode`, by changing the endpoint selector? So publisher doesn't need to know if endpoints are eligible or not.

If you prefer to have "HealtCheck" and "Probe" annotation linked to an endpoint which have to be (maybe) manually ignored by each publisher, it is correct that endpoint still be available in the list of the endpoints of a resource? e.g. here https://github.com/dotnet/aspire/blob/0cd12b32d3d177cdaf2a508a5a99a15cd4776581/src/Aspire.Hosting.Azure.AppContainers/ContainerAppContext.cs#L222 or it has to be ignored?

Honestly I'm missing something, where is the benefit of have an endpoint automatically added (especially in non run mode) but (maybe) ignored and (maybe) potentially not handled by the service itself?

Dona278 committed:
Merge branch 'main' into probes

Dona278 committed:
Merge branch 'main' into probes

Dona278 committed:
Moved "AddProbes" to "BaseContainerAppContext"

davidfowl commented on src/Aspire.Hosting/ApplicationModel/ProbeAnnotation.cs:
```diff
@@ -0,0 +1,77 @@
+
+/// <summary>
+/// Enum representing the type of probe.
+/// </summary>
```
Mark these all as experimental.

Dona278 committed:
Marked "ProbeAnnotation" as experimental

Dona278 committed:
Forced "http" scheme for probe which have associated an endpoint used for ingress

Dona278 committed:
Merge branch 'main' into probes

davidfowl commented:
Verified ACA and AppService, but not k8s. This API is fully experimental and we will have a chance to tweak it as break it as we learn more.
